samples/oauth_rails_vethealth/app/controllers/session_controller.rb

```diff
@@ -0,0 +1,125 @@
+      {param: :response_type, description: "Set to <code>code</code> to use the Authorization Code Flow.", required: true},
+      {param: :scope, description: "The information for which your app is requesting access.  Should include <code>openid</code>.  <code>launch/patient</code> is used to get the patient id to use as a param for Health API requests.  <code>profile</code> returns user information like name and email.  The <code>patient/*</code> scopes give access to each of the specific APIs".html_safe},
+      {param: :state, description: "Can be used as a nonce for security or as app state information."}
+    ]
+    @oauth_url = "https://dev-api.va.gov/oauth2/authorization?#{@oauth_params.to_query}"
+  end
+
+  def callback
+    # display content of the callback, post to oauth, and resulting data (including token) received from oauth
+    # this controller is heavier than a typical callback enpdpoint would be to support showing all that data (as well as supporting refreshing the page), for another callback example see https://github.com/department-of-veterans-affairs/vets-api-clients/tree/master/samples/oauth_rails_vetverification
+    if params[:code].nil?
+      if params[:error].present?
+        flash.alert = "Callback received \"#{params['error']}\" error: #{params['error_description']}"
+      else
+        flash.notice = "The callback page is for oauth responses, please login first."
+      end
+      redirect_to(login_path) and return
+    end
+    @verified_state = params[:state].to_i == session[:login_time]
+    @body = {
+      code: params[:code],
+      state: params[:state],
+      grant_type: 'authorization_code',
+      redirect_uri: 'http://localhost:3000/callback'
+    }
+    @post_url = 'https://dev-api.va.gov/oauth2/token'
+    auth = { username: ENV['va_developer_client_id'], password: ENV['va_developer_client_secret'] }
+    @auth = auth.dup
+    @auth[:password] = "<Client Secret found in application.yml>"
+    @session = nil
+
+    # set oauth_response and oauth_response_code
+    if session[:oauth_code]
+      # we are reloading the page so we don't want to make a stale request
+      @oauth_response_code = session[:oauth_code]
+      @oauth_response = 
+        if session[:oauth_response].is_a? Hash
+          # response was not saved as a Session model
+          if @oauth_response_code == 200
+            # 200 OK means it was a successful request but session was found invalid
+            attributes = Session.attributes_from_oauth(oauth_post)
+            @session = Session.new(attributes)
+          end
+          session[:oauth_response]
+        else
+          # session[:oauth_response] is an array of keys, so get data from saved session
+          @session = Session.find(session[:id])
+          hide_sensitive_data(@session.attributes.reject { |k,v| %w(id created_at updated_at).include?(k) })
+        end
+      @session.validate_session(session) if @session
```

Comment:
I am pretty confused about what the difference is between `@session` and `session` is, so this line hurt my head a little

Reply:
renamed `Session` model to `Authentication`